src/libraries/System.Net.Http/tests/UnitTests/Headers/ObjectCollectionTest.cs

```diff
@@ -36,9 +36,29 @@ public void Ctor_ExecuteBothOverloads_MatchExpectation()
             c.Add("value1");
 
             Assert.Throws<InvalidOperationException>(() => { c.Add(null); });
+        }
+
+        [Fact]
+        public void ContainsAndRemove_UsesEqualitySemantics()
+        {
+            // Use default validator
+            ObjectCollection<string> c = new ObjectCollection<string>();
+
+            Assert.Equal(0, c.Count);
+            Assert.False(c.Contains("value" + 1));
+
+            c.Add("value" + 1);
 
             Assert.Equal(1, c.Count);
-            Assert.True(c.Contains("value1"));
+            // Force the reference to be different to ensure we are checking for semantic equality
+            // and not reference equality.
+            Assert.True(c.Contains("value" + 1));
+            c.Add("value" + 2);
+
+            Assert.True(c.Remove("value" + 1));
+            Assert.Equal(1, c.Count);
+
+            Assert.True(c.Contains("value" + 2));
```

Comment:
I suggest we put some of these values being looked up into locals so that it's really obvious why the scheme is being used.  We should also validate the removal path when a single item is stored, e.g.
```C#
string value1 = "value1";
string value1DifferentReference = "value" + 1;
Assert.NotSame(value1, value1DifferentReference);
Assert.Equal(value1, value1DifferentReference);

string value2 = "value2";
string value2DifferentReference = "value" + 2;
Assert.NotSame(value1, value2DifferentReference);
Assert.Equal(value1, value2DifferentReference);

string value3 = "value3";

// Empty collection
Assert.Equal(0, c.Count);
Assert.False(c.Contains(value1));

// One item
c.Add(value1);
Assert.Equal(1, c.Count);
Assert.True(c.Contains(value1));
Assert.True(c.Contains(value1DifferentReference));
Assert.False(c.Contains(value2));

// Remove it
Assert.True(c.Remove(value1DifferentReference));
Assert.Equal(0, c.Count);
Assert.False(c.Contains(value1));

// Two items
c.Add(value1);
c.Add(value2);
Assert.Equal(2, c.Count);
Assert.True(c.Contains(value1));
Assert.True(c.Contains(value1DifferentReference));
Assert.True(c.Contains(value2));
Assert.True(c.Contains(value2DifferentReference));
Assert.False(c.Contains(value3));

// Remove one
Assert.True(c.Remove(value1DifferentReference));
Assert.Equal(1, c.Count);
Assert.False(c.Contains(value1));
Assert.True(c.Contains(value2));

// Remove the other
Assert.True(c.Remove(value2DifferentReference));
Assert.Equal(0, c.Count);
Assert.False(c.Contains(value1));
Assert.False(c.Contains(value2));

// Remove something that doesn't exist
Assert.False(c.Remove(value1));
Assert.False(c.Remove(value2));
Assert.False(c.Remove(value3));
```